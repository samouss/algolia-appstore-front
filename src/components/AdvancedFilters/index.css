.AdvancedFilters {
  display: flex;
  flex-direction: column;
  margin-top: 1vh;
}

.AdvancedFilters__Content {
  align-items: flex-start;
  margin-bottom: 1vh;
  display: none;

  @media (--tablet) {
    display: flex;
    flex-direction: row;
    margin-bottom: 0;
  }
}

.AdvancedFilters__Content--visible {
  display: flex;
  flex-direction: column;
}

.AdvancedFilters__Link {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  min-height: 25px;
  color: #b9b9b9;
  font-size: 0.8rem;

  @media (--tablet) {
    display: none;
  }
}
