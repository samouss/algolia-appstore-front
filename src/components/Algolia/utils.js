import escape from 'lodash.escape';

export const highlightPreTag = '__algolia__highlight__pre__tag__';
export const highlightPostTag = '__algolia__highlight__post__tag__';

export const escapeAndReplaceTag = value => {
  return escape(value)
    .replace(highlightPreTag, '<em>')
    .replace(highlightPostTag, '</em>');
};
