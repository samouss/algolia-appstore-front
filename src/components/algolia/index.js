export { default as Provider } from './Provider';
export { default as withPaginateResults } from './withPaginateResults';
