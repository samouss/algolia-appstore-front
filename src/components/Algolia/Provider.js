import { Component, Children } from 'react';
import PropTypes from 'prop-types';
import algoliasearch from 'algoliasearch';
import algoliasearchHelper from 'algoliasearch-helper';
import { highlightPreTag, highlightPostTag } from './utils';

export const ContextTypes = {
  algoliaClient: PropTypes.object.isRequired,
  algoliaHelper: PropTypes.object.isRequired,
};

class Provider extends Component {
  constructor(props) {
    super(props);

    this.client = algoliasearch(props.appId, props.apiKey);
    this.helper = algoliasearchHelper(this.client, props.indexName, props.options);
  }

  getChildContext() {
    return {
      algoliaClient: this.client,
      algoliaHelper: this.helper,
    };
  }

  componentDidMount() {
    this.helper
      .setQueryParameter('highlightPreTag', highlightPreTag)
      .setQueryParameter('highlightPostTag', highlightPostTag)
      .search();
  }

  render() {
    return Children.only(this.props.children);
  }
}

Provider.propTypes = {
  appId: PropTypes.string.isRequired,
  apiKey: PropTypes.string.isRequired,
  indexName: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  options: PropTypes.object,
};

Provider.defaultProps = {
  options: {},
};

Provider.childContextTypes = ContextTypes;

export default Provider;
