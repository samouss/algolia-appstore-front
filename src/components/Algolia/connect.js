import React from 'react';
import PropTypes from 'prop-types';
import { getDisplayName } from 'core/utils';
import { ContextTypes } from './Provider';

export const ConnectPropTypes = {
  helper: PropTypes.object.isRequired,
};

export const connect = WrappedComponent => {
  const Connect = (props, context) => (
    <WrappedComponent
      {...props}
      helper={context.algoliaHelper}
    />
  );

  Connect.displayName = getDisplayName(
    WrappedComponent,
    'connect',
  );

  Connect.contextTypes = ContextTypes;

  return Connect;
};

export default connect;
