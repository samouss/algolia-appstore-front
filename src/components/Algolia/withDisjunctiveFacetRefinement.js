import React, { Component } from 'react';
import { getDisplayName } from 'core/utils';
import { ContextTypes } from './Provider';

const withDisjunctiveFacetRefinement = ({
  facet,
} = {}) => WrappedComponent => {
  class WithDisjunctiveFacetRefinement extends Component {

    constructor(props) {
      super(props);

      this.onChange = this.onChange.bind(this);
    }

    onChange(facetValue, isRefined) {
      const { algoliaHelper } = this.context;

      if (isRefined) {
        return algoliaHelper
          .removeDisjunctiveFacetRefinement(facet)
          .search();
      }

      algoliaHelper
        .removeDisjunctiveFacetRefinement(facet)
        .addDisjunctiveFacetRefinement(facet, facetValue)
        .search();
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          onChange={this.onChange}
        />
      );
    }

  }

  WithDisjunctiveFacetRefinement.displayName = getDisplayName(
    WrappedComponent,
    'withDisjunctiveFacetRefinement',
  );

  WithDisjunctiveFacetRefinement.contextTypes = ContextTypes;

  return WithDisjunctiveFacetRefinement;
};

export default withDisjunctiveFacetRefinement;
