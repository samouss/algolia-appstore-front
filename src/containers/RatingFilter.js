import flowRight from 'lodash.flowright';
import { withFacet, withNumericRefinement } from 'components/Algolia';
import RatingList from 'components/RatingList';

const facet = 'rating';
const operator = '>=';

export const reduceFacetValues = (facetValues, helperState) => {
  const currentRefinementsForOperator = helperState.getNumericRefinement(facet, operator) || [];

  return [5, 4, 3, 2, 1, 0].reduce((acc, rating) => {
    const countCurrentRating = facetValues
      .filter(_ => parseInt(_.name, 10) >= rating)
      .reduce((acc, facetValue) => acc + facetValue.count, 0);

    const isRatingRefined = currentRefinementsForOperator.includes(rating);
    const isCountLessThanZero = countCurrentRating <= 0;
    const isFacetValuesEmpty = !facetValues.length;

    if (isFacetValuesEmpty || (!isRatingRefined && isCountLessThanZero)) {
      return acc;
    }

    return acc.concat({
      name: rating.toString(),
      count: countCurrentRating,
      isRefined: isRatingRefined,
    });
  }, []);
};

export default flowRight(
  withFacet({
    facet,
    reduceFacetValues,
  }),
  withNumericRefinement({
    attribute: facet,
    operator,
  }),
)(RatingList);
