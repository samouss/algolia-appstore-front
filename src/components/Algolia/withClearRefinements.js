import flowRight from 'lodash.flowright';
import React, { Component } from 'react';
import { getDisplayName } from 'core/utils';
import connect, { ConnectPropTypes } from './connect';

export const withClearRefinements = WrappedComponent => {
  class WithClearRefinements extends Component {

    constructor(props) {
      super(props);

      this.onClear = this.onClear.bind(this);
    }

    onClear() {
      const { helper } = this.props;

      helper
        .clearRefinements()
        .search();
    }

    render() {
      const { helper, ...props } = this.props;

      return (
        <WrappedComponent
          {...props}
          onClear={this.onClear}
        />
      );
    }

  }

  WithClearRefinements.displayName = getDisplayName(
    WrappedComponent,
    'withClearRefinements',
  );

  WithClearRefinements.propTypes = ConnectPropTypes;

  return WithClearRefinements;
};

export default flowRight(
  connect,
  withClearRefinements,
);
