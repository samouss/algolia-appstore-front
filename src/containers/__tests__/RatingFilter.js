import React from 'react';
import { shallow } from 'enzyme';
import { withFacet, withNumericRefinement } from 'components/Algolia';
import RatingFilter, { reduceFacetValues } from '../RatingFilter';

jest.mock('components/Algolia', () => ({
  withFacet: jest.fn(() => Component => props => (
    <Component
      {...props}
      facetValues={[]}
    />
  )),
  withNumericRefinement: jest.fn(() => Component => props => (
    <Component
      {...props}
      onChange={() => {}}
    />
  )),
}));

describe('<RatingFilter />', () => {
  it('expect to render', () => {
    const component = shallow(<RatingFilter />).shallow();

    expect(component).toMatchSnapshot();
  });

  describe('withFacet', () => {
    it('expect to be called', () => {
      expect(withFacet).toHaveBeenCalledWith({
        facet: 'rating',
        reduceFacetValues,
      });
    });
  });

  describe('withDisjunctiveFacetRefinement', () => {
    it('expect to be called', () => {
      expect(withNumericRefinement).toHaveBeenCalledWith({
        attribute: 'rating',
        operator: '>=',
      });
    });
  });

  describe('reduceFacetValues', () => {
    const helperState = {
      getNumericRefinement: jest.fn(),
    };

    it('expect to return facetValues grouped by 5 to 0 without refined', () => {
      const facetValues = [
        { name: '5', count: 1 },
        { name: '4.5', count: 1 },
        { name: '4', count: 1 },
        { name: '3.5', count: 1 },
        { name: '3', count: 1 },
        { name: '2.5', count: 1 },
        { name: '2', count: 1 },
        { name: '1.5', count: 1 },
        { name: '1', count: 1 },
      ];

      const expectation = [
        { name: '5', count: 1, isRefined: false },
        { name: '4', count: 3, isRefined: false },
        { name: '3', count: 5, isRefined: false },
        { name: '2', count: 7, isRefined: false },
        { name: '1', count: 9, isRefined: false },
        { name: '0', count: 9, isRefined: false },
      ];

      const actual = reduceFacetValues(facetValues, helperState);

      expect(actual).toEqual(expectation);
    });

    it('expect to return facetValues grouped by 5 to 0 without zero count', () => {
      const facetValues = [
        { name: '3.5', count: 1 },
        { name: '3', count: 1 },
        { name: '2.5', count: 1 },
        { name: '2', count: 1 },
      ];

      const expectation = [
        { name: '3', count: 2, isRefined: false },
        { name: '2', count: 4, isRefined: false },
        { name: '1', count: 4, isRefined: false },
        { name: '0', count: 4, isRefined: false },
      ];

      const actual = reduceFacetValues(facetValues, helperState);

      expect(actual).toEqual(expectation);
    });

    it('expect to return facetValues grouped by 5 to 0 with refined', () => {
      helperState.getNumericRefinement.mockImplementationOnce(() => [2]);

      const facetValues = [
        { name: '5', count: 1 },
        { name: '4.5', count: 1 },
        { name: '4', count: 1 },
        { name: '3.5', count: 1 },
        { name: '3', count: 1 },
        { name: '2.5', count: 1 },
        { name: '2', count: 1 },
        { name: '1.5', count: 1 },
        { name: '1', count: 1 },
      ];

      const expectation = [
        { name: '5', count: 1, isRefined: false },
        { name: '4', count: 3, isRefined: false },
        { name: '3', count: 5, isRefined: false },
        { name: '2', count: 7, isRefined: true },
        { name: '1', count: 9, isRefined: false },
        { name: '0', count: 9, isRefined: false },
      ];

      const actual = reduceFacetValues(facetValues, helperState);

      expect(actual).toEqual(expectation);
    });

    it('expect to return facetValues grouped by 5 to 0 with refined and zero count', () => {
      helperState.getNumericRefinement.mockImplementationOnce(() => [4]);

      const facetValues = [
        { name: '3.5', count: 1 },
        { name: '3', count: 1 },
        { name: '2.5', count: 1 },
        { name: '2', count: 1 },
      ];

      const expectation = [
        { name: '4', count: 0, isRefined: true },
        { name: '3', count: 2, isRefined: false },
        { name: '2', count: 4, isRefined: false },
        { name: '1', count: 4, isRefined: false },
        { name: '0', count: 4, isRefined: false },
      ];

      const actual = reduceFacetValues(facetValues, helperState);

      expect(actual).toEqual(expectation);
    });

    it('expect to return empty array when facetValues is empty', () => {
      const facetValues = [];

      const expectation = [];
      const actual = reduceFacetValues(facetValues, helperState);

      expect(actual).toEqual(expectation);
    });
  });
});
