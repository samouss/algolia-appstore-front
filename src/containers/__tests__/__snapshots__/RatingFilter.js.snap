// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RatingFilter /> expect to render 1`] = `
<RatingList
  facetValues={Array []}
  onChange={[Function]}
/>
`;
